src/hardware/vmware.cpp

```diff
@@ -0,0 +1,156 @@
+
+// Magic number for all VMware calls
+static constexpr uint32_t VMWARE_MAGIC = 0x564d5868u;
+
+enum class VmWareCommand : uint16_t {
+	GetVersion        = 10,
+	AbsPointerData    = 39,
+	AbsPointerStatus  = 40,
+	AbsPointerCommand = 41,
+};
+
+enum class VmWarePointer : uint32_t {
+	Enable   = 0x45414552,
+	Relative = 0xf5,
+	Absolute = 0x53424152,
+};
+
+// ***************************************************************************
+// Request handling
+// ***************************************************************************
+
+static void command_get_version()
+{
+	reg_eax = 0; // protocol version
+	reg_ebx = VMWARE_MAGIC;
+}
+
+static void command_abs_pointer_data()
+{
+	MouseVmWarePointerStatus status = {};
+	MOUSEVMM_GetPointerStatus(status);
+
+	reg_eax = status.buttons;
+	reg_ebx = status.absolute_x;
+	reg_ecx = status.absolute_y;
+	reg_edx = status.wheel_counter;
+}
+
+static void command_abs_pointer_status()
+{
+	constexpr uint32_t ABS_UPDATED     = 4;
+	constexpr uint32_t ABS_NOT_UPDATED = 0;
+
+	reg_eax = MOUSEVMM_CheckIfUpdated_VmWare() ? ABS_UPDATED : ABS_NOT_UPDATED;
+}
+
+static void command_abs_pointer()
+{
+	switch (static_cast<VmWarePointer>(reg_ebx)) {
+	case VmWarePointer::Enable: break; // can be safely ignored
+	case VmWarePointer::Relative:
+		MOUSEVMM_Deactivate(MouseVmmProtocol::VmWare);
+		break;
+	case VmWarePointer::Absolute:
+		MOUSEVMM_Activate(MouseVmmProtocol::VmWare);
+		break;
+	default:
+		LOG_WARNING("VMWARE: unimplemented mouse subcommand 0x%08x", reg_ebx);
+		break;
+	}
+}
+
+// ***************************************************************************
+// I/O port
+// ***************************************************************************
+
+static uint32_t port_read_vmware(const io_port_t, const io_width_t)
+{
+	if (reg_eax != VMWARE_MAGIC) {
+		return 0;
+	}
+
+	switch (static_cast<VmWareCommand>(reg_cx)) {
+	case VmWareCommand::GetVersion: command_get_version(); break;
+	case VmWareCommand::AbsPointerData: command_abs_pointer_data(); break;
+	case VmWareCommand::AbsPointerStatus: command_abs_pointer_status(); break;
+	case VmWareCommand::AbsPointerCommand: command_abs_pointer(); break;
+	default:
+		LOG_WARNING("VMWARE: unimplemented command 0x%08x", reg_ecx);
+		break;
+	}
+
+	return reg_eax;
+}
+
+// ***************************************************************************
+// Lifecycle
+// ***************************************************************************
+
+void VMWARE_Destroy(Section*)
+{
+	if (is_interface_enabled) {
+		IO_FreeReadHandler(port_num_virtualbox, io_width_t::dword);
+		is_interface_enabled = false;
+	}
+}
+
+void VMWARE_Init(Section* sec)
+{
+	has_feature_mouse = MOUSEVMM_IsSupported(MouseVmmProtocol::VmWare);
+
+	// TODO: implement more features:
+	// - shared directories, for VMSMount tool:
+	//   https://github.com/eduardocasino/vmsmount
+	// - everything supported by Windows 9x Guest Additions
```

Comment:
I think the proper name is "Windows 9x VMware Tools", because Guest Additions is a VirtualBox term?

Reply:
OK, changed.